### PR TITLE
[Api2.0] add pixel shuffle class

### PR DESCRIPTION
**Pull request by shippingwang**
### PR types

New features

### PR changes

APIs

### Describe

This PR
- add pixel shuffle functional function and PixelShuffle class 
- support data_format in pixel_shuffle api

### CODE REVIEW

File: python/paddle/nn/functional/vision.py [hunk 1]
```diff
@@ -12,80 +12,113 @@
...
 __all__ = [
-    'affine_channel',
-    'affine_grid',
-    'anchor_generator',
-    'bipartite_match',
-    'box_clip',
-    'box_coder',
-    'box_decoder_and_assign',
-    'collect_fpn_proposals',
+    #'affine_channel',
```
Comment (willthefrog) [inline python/paddle/nn/functional/vision.py]: why are these commented out?

File: python/paddle/nn/functional/__init__.py [hunk 2]
```diff
@@ -158,40 +158,40 @@
...
 # from .rnn import lstm_unit        #DEFINE_ALIAS
-from .vision import affine_channel  #DEFINE_ALIAS
-from .vision import affine_grid  #DEFINE_ALIAS
-from .vision import anchor_generator  #DEFINE_ALIAS
-from .vision import bipartite_match  #DEFINE_ALIAS
-from .vision import box_clip  #DEFINE_ALIAS
-from .vision import box_coder  #DEFINE_ALIAS
-from .vision import box_decoder_and_assign  #DEFINE_ALIAS
-from .vision import collect_fpn_proposals  #DEFINE_ALIAS
+#from .vision import affine_channel  #DEFINE_ALIAS
```
Comment (willthefrog) [inline python/paddle/nn/functional/__init__.py]: why are these commented out?

File: python/paddle/nn/functional/vision.py [hunk 3]
```diff
@@ -12,80 +12,113 @@
...
+    #'shuffle_channel',
+    #'space_to_depth',
+    #'yolo_box',
+    #'yolov3_loss'
 ]
+
+
+def pixel_shuffle(x, upscale_factor, data_format="NCHW", name=None):
+    """
+        :alias_main: paddle.nn.functional.pixel_shuffle
```
Comment (willthefrog) [inline python/paddle/nn/functional/vision.py]: remove alias

File: python/paddle/nn/functional/vision.py [hunk 4]
```diff
@@ -12,80 +12,113 @@
...
+        :alias: paddle.nn.functional.pixel_shuffle,paddle.nn.functional.vision.pixel_shuffle
+    
+    This operator rearranges elements in a tensor of shape [N, C, H, W]
+    to a tensor of shape [N, C/upscale_factor**2, H*upscale_factor, W*upscale_factor],
+    or from shape [N, H, W, C] to [N, H*upscale_factor, W*upscale_factor, C/upscale_factor^2].
```
Comment (willthefrog) [inline python/paddle/nn/functional/vision.py]: use the same "** 2" notation?

File: python/paddle/nn/functional/vision.py [hunk 5]
```diff
@@ -12,80 +12,113 @@
...
+    This is useful for implementing efficient sub-pixel convolution
+    with a stride of 1/upscale_factor.
+    Please refer to the paper: `Real-Time Single Image and Video Super-Resolution
+    Using an Efficient Sub-Pixel Convolutional Neural Network <https://arxiv.org/abs/1609.05158v2>`_ .
+    by Shi et. al (2016) for more details.
+
+    Parameters:
+
+        x(Variable): 4-D tensor, the data type should be float32 or float64.
```
Comment (willthefrog) [inline python/paddle/nn/functional/vision.py]: `variable` -> `tensor`

File: python/paddle/nn/functional/vision.py [hunk 6]
```diff
@@ -12,80 +12,113 @@
...
+            paddle.disable_static()
+
+            x_var = paddle.to_variable(x)
+            out_var = F.pixel_shuffle(x_var, 3)
+            out = out_var.numpy()
+            print(out.shape) 
+            # (2, 1, 12, 12)
+
+    """
+    if not in_dygraph_mode():
```
Comment (willthefrog) [inline python/paddle/nn/functional/vision.py:289]: add fast path for dygraph mode, see the guideline

File: python/paddle/nn/functional/vision.py [hunk 7]
```diff
@@ -12,80 +12,113 @@
...
+        .. code-block:: python
+
+            import paddle
+            import paddle.nn.functional as F
+            import numpy as np
+
+            x = np.random.randn(2, 9, 4, 4).astype(np.float32)
+            paddle.disable_static()
+
+            x_var = paddle.to_variable(x)
```
Comment (willthefrog) [inline python/paddle/nn/functional/vision.py]: `to_tensor`

File: python/paddle/nn/layer/vision.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+
+from ...fluid.dygraph import layers
+from .. import functional
+
+__all__ = ['PixelShuffle']
+
+
+class PixelShuffle(layers.Layer):
+    """
+        :alias_main: paddle.nn.PixelShuffle
```
Comment (willthefrog) [inline python/paddle/nn/layer/vision.py]: remove alias

File: python/paddle/nn/layer/vision.py [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+        :alias: paddle.nn.PixelShuffle,paddle.nn.layer.PixelShuffle,paddle.nn.layer.vision.PixelShuffle
+    
+    PixelShuffle Layer    
+
+    This operator rearranges elements in a tensor of shape [N, C, H, W]
+    to a tensor of shape [N, C/upscale_factor**2, H*upscale_factor, W*upscale_factor],
+    or from shape [N, H, W, C] to [N, H*upscale_factor, W*upscale_factor, C/upscale_factor^2].
```
Comment (willthefrog) [inline python/paddle/nn/layer/vision.py]: "**2" notation.

File: python/paddle/nn/functional/vision.py [hunk 10]
```diff
@@ -12,80 +12,113 @@
...
+    #'roi_pool',
+    #'shuffle_channel',
+    #'space_to_depth',
+    #'yolo_box',
+    #'yolov3_loss'
 ]
+
+
+def pixel_shuffle(x, upscale_factor, data_format="NCHW", name=None):
+    """
```
Comment (willthefrog) [inline python/paddle/nn/functional/vision.py]: no need for duplicated docstrings for functional version, just refer to class doc

File: python/paddle/nn/layer/vision.py [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+    This is useful for implementing efficient sub-pixel convolution
+    with a stride of 1/upscale_factor.
+    Please refer to the paper: `Real-Time Single Image and Video Super-Resolution
+    Using an Efficient Sub-Pixel Convolutional Neural Network <https://arxiv.org/abs/1609.05158v2>`_ .
+    by Shi et. al (2016) for more details.
+
+    Parameters:
+
+        upscale_factor(int): factor to increase spatial resolution.
```
Comment (willthefrog) [inline python/paddle/nn/layer/vision.py:39]: missing `name`

File: paddle/fluid/operators/pixel_shuffle_op.cc [hunk 12]
```diff
@@ -70,6 +89,12 @@ class PixelShuffleOpMaker : public framework::OpProtoAndCheckerMaker {
                             platform::errors::InvalidArgument(
                                 "upscale_factor should be larger than 0."));
         });
+    AddAttr<std::string>(
+        "data_format",
+        "(string, default NCHW) Only used in "
```
Comment (willthefrog) [inline paddle/fluid/operators/pixel_shuffle_op.cc]: delete this line?

**Review by willthefrog — CHANGES_REQUESTED** [review]

File: python/paddle/nn/functional/vision.py [hunk 13]
```diff
@@ -89,3 +92,68 @@
...
+    or from shape [N, H, W, C] to [N, H*upscale_factor, W*upscale_factor, C/upscale_factor**2].
+    This is useful for implementing efficient sub-pixel convolution
+    with a stride of 1/upscale_factor.
+    Please refer to the paper: `Real-Time Single Image and Video Super-Resolution
+    Using an Efficient Sub-Pixel Convolutional Neural Network <https://arxiv.org/abs/1609.05158v2>`_ .
+    by Shi et. al (2016) for more details.
+
+    Parameters:
+
+        x(tensor): 4-D tensor, the data type should be float32 or float64.
```
Comment (willthefrog) [inline python/paddle/nn/functional/vision.py]: `(tensor)` -> `(Tensor)`

File: python/paddle/nn/functional/vision.py [hunk 14]
```diff
@@ -89,3 +92,68 @@
...
+        upscale_factor(int): factor to increase spatial resolution.
+        name (str, optional): The default value is None.  Normally there is no need for user to set this property.
+
+    Returns:
+        Out(Variable): Reshaped tensor according to the new dimension.
```
Comment (willthefrog) [inline python/paddle/nn/functional/vision.py]: `variable` -> `tensor`

**Review by willthefrog — CHANGES_REQUESTED** [review]

**Comment by shippingwang** [comment]
@willthefrog Codes have been refined after first time reviewing, feel free to re-review while passing tests

**Review by willthefrog — CHANGES_REQUESTED** [review]
please reply to each comment  before making new review request, otherwise it is hard  for the reviewer to keep track of the changes

File: python/paddle/nn/functional/vision.py (hunk 1 above)
Comment (shippingwang) [reply python/paddle/nn/functional/vision.py]: already restored it.

File: python/paddle/nn/functional/__init__.py (hunk 2 above)
Comment (shippingwang) [reply python/paddle/nn/functional/__init__.py]: Already restored it.

File: python/paddle/nn/functional/vision.py (hunk 3 above)
Comment (shippingwang) [reply python/paddle/nn/functional/vision.py]: Done

File: python/paddle/nn/functional/vision.py (hunk 5 above)
Comment (shippingwang) [reply python/paddle/nn/functional/vision.py]: Done

File: python/paddle/nn/functional/vision.py (hunk 6 above)
Comment (shippingwang) [reply python/paddle/nn/functional/vision.py:289]: Done. Follow the guide, ```core.ops.XXX``` has been added in dygraph mode, but it confuses me, why does it follow this format: <Tensor, String(Attr name), Attr>, It looks like the Attr name is redundancy if the params list has an order, or it is just the Pybind rule?

File: python/paddle/nn/functional/vision.py (hunk 7 above)
Comment (shippingwang) [reply python/paddle/nn/functional/vision.py]: Done

File: python/paddle/nn/layer/vision.py (hunk 8 above)
Comment (shippingwang) [reply python/paddle/nn/layer/vision.py]: Done

File: python/paddle/nn/layer/vision.py (hunk 9 above)
Comment (shippingwang) [reply python/paddle/nn/layer/vision.py]: Done

File: python/paddle/nn/functional/vision.py (hunk 10 above)
Comment (shippingwang) [reply python/paddle/nn/functional/vision.py]: Yep, It is the same thing.

File: paddle/fluid/operators/pixel_shuffle_op.cc (hunk 12 above)
Comment (shippingwang) [reply paddle/fluid/operators/pixel_shuffle_op.cc]: Done

File: python/paddle/nn/functional/vision.py (hunk 13 above)
Comment (shippingwang) [reply python/paddle/nn/functional/vision.py]: Done

File: python/paddle/nn/functional/vision.py (hunk 14 above)
Comment (shippingwang) [reply python/paddle/nn/functional/vision.py]: Done

**Comment by shippingwang** [comment]
> please reply to each comment before making new review request, otherwise it is hard for the reviewer to keep track of the changes

Done

File: python/paddle/nn/functional/vision.py (hunk 10 above)
Comment (willthefrog) [reply python/paddle/nn/functional/vision.py]: literally, just remove the duplicated description (bug keep the args and stuff) and  add a reference to the Layer version.

**Review by willthefrog — CHANGES_REQUESTED** [review]

**Review by willthefrog — DISMISSED** [review]
LGTM

**Review by willthefrog — APPROVED** [review]
LGTM
